crates/ra_ide/src/syntax_highlighting.rs

```diff
@@ -55,13 +55,57 @@ pub(crate) fn highlight(
     };
 
     let mut bindings_shadow_count: FxHashMap<Name, u32> = FxHashMap::default();
-    let mut res = Vec::new();
+    // We use a stack for the DFS traversal below.
+    // When we leave a node, the we use it to flatten the highlighted ranges.
+    let mut res: Vec<Vec<HighlightedRange>> = vec![Vec::new()];
 
     let mut current_macro_call: Option<ast::MacroCall> = None;
 
     // Walk all nodes, keeping track of whether we are inside a macro or not.
     // If in macro, expand it first and highlight the expanded code.
     for event in root.preorder_with_tokens() {
+        assert!(!res.is_empty());
```

Comment:
Should we have this `assert!` here?

Reply:
The invariant is that the stack never becomes empty. This is asserted before and after the modification. I am fine with removing this assertion, I just usually err on the side of too many assertions 😉